salt/_utils/kubernetes_utils.py

```diff
@@ -0,0 +1,535 @@
+        name='namespaced_role_binding',
+    ),
+    # }}}
+    # /apis/storage.k8s.io/v1/ {{{
+    ('storage.k8s.io/v1', 'StorageClass'): ObjectDetails(
+        model=k8s_client.V1StorageClass,
+        api=k8s_client.StorageV1Api,
+        name='storage_class',
+    ),
+    # }}}
+}
+
+
+# CustomResources cannot rely on statically declared models, which is why their
+# management is treated differently from "standard" objects.
+# We define a `CRDetails` object similar to `ObjectDetails`, and provide a
+# simple `CustomObject` to emulate a "standard" model.
+class CRDetails:
+    def __init__(self, api_version, kind, scope, plural):
+        group, _, version = api_version.rpartition('/')
+        if not (group and version):
+            raise ValueError(
+                "Malformed 'apiVersion': {} "
+                "(expected format '<group>/<version>')".format(api_version)
+            )
+
+        self._group = group
+        self._version = version
+        self._kind = kind
+        scope = ObjectScope(scope)
+        self._scope = scope
+        self._plural = plural
+
+        self._create = getattr(CustomObjectsApi, self._method('create'))
+        self._retrieve = getattr(CustomObjectsApi, self._method('get'))
+        self._update = getattr(CustomObjectsApi, self._method('patch'))
+        self._delete = getattr(CustomObjectsApi, self._method('delete'))
+        self._replace = getattr(CustomObjectsApi, self._method('replace'))
+
+    group = property(operator.attrgetter('_group'))
+    version = property(operator.attrgetter('_version'))
+    kind = property(operator.attrgetter('_kind'))
+    scope = property(operator.attrgetter('_scope'))
+    plural = property(operator.attrgetter('_plural'))
+    create = property(operator.attrgetter('_create'))
+    retrieve = property(operator.attrgetter('_retrieve'))
+    update = property(operator.attrgetter('_update'))
+    delete = property(operator.attrgetter('_delete'))
+    replace = property(operator.attrgetter('_replace'))
+
+    @property
+    def api_version(self):
+        return '/'.join([self.group, self.version])
+
+    @property
+    def key(self):
+        # Used for indexing in `KNOWN_CUSTOM_OBJECTS`
+        return (self.api_version, self.kind)
+
+    def _method(self, verb):
+        return '{}_{}_custom_object'.format(verb, self.scope.value)
+
+
+_CUSTOM_OBJECTS = [
+    CRDetails('monitoring.coreos.com/v1', 'Alertmanager',
+              scope='namespaced', plural='alertmanagers'),
+    CRDetails('monitoring.coreos.com/v1', 'Prometheus',
+              scope='namespaced', plural='prometheuses'),
+    CRDetails('monitoring.coreos.com/v1', 'PrometheusRule',
+              scope='namespaced', plural='prometheusrules'),
+    CRDetails('monitoring.coreos.com/v1', 'ServiceMonitor',
+              scope='namespaced', plural='servicemonitors'),
+]
+
+KNOWN_CUSTOM_OBJECTS = {details.key: details for details in _CUSTOM_OBJECTS}
+
+
+class _DictWrapper(object):
```

Comment:
This needs some more documentation (also `CustomObject`) IMHO as to what the intent is and how it's achieved. Not immediately obvious.

Reply:
OK, will try to give more details.